apps/consulting/posts/review-functorch-contribution.md

```diff
@@ -0,0 +1,327 @@
+our function didn't have side effects and graph-breaks. Thus, we were able to
+successfully trace through this program in one graph.
+
+```python
+class GraphModule(torch.nn.Module):
+    def forward(self, L_x_ : torch.Tensor):
+        l_x_ = L_x_
+        
+        # File: torch/_functorch/apis.py:363, code:
+        # return eager_transforms.grad_impl(func, argnums, has_aux, args, kwargs)
+        grad_body_0 = self.grad_body_0
+        grad_proxy = torch.func.grad(grad_body_0, 0, False);  grad_body_0 = None
+        call = grad_proxy.__call__(l_x_);  grad_proxy = l_x_ = None
+        contiguous = call.contiguous();  call = None
+        return (contiguous,)
+        
+    class GraphModule(torch.nn.Module):
+        def forward(self, l_x_):
+            # No stacktrace found for following nodes
+            _set_grad_enabled = torch._C._set_grad_enabled(True)
+
+            # File: test/test_scratch.py:382, code: return torch.sin(x)
+            sin = torch.sin(l_x_);  l_x_ = None
+
+            # No stacktrace found for following nodes
+            _set_grad_enabled_1 = torch._C._set_grad_enabled(True)
+            return sin
+```
+
+The graph shown above is handed over to `aot_autograd` for the subsequent phase
+of the compilation process. `aot_autograd` performs a trace through the
+transformation, resulting in the generation of the transformed graph. This
+explains why we observe a call to `cos` instead of `sin`, as the transformation
+has effectively altered the operation from what was originally present in our function.
+
+```python
+def forward(self, arg0_1: f32[]):
+    # File: torch/_functorch/apis.py:363,
+    # code: return eager_transforms.grad_impl(func, argnums, has_aux, args, kwargs)
+    full: f32[] = torch.ops.aten.full.default([], 1, dtype = torch.float32,
+                                              layout = torch.strided,
+                                              device = device(type='cpu'),
+                                              pin_memory = False)
+    cos: f32[] = torch.ops.aten.cos.default(arg0_1);  arg0_1 = None
+    mul: f32[] = torch.ops.aten.mul.Tensor(full, cos);  full = cos = None
+    return (mul,)
+```
+
+The inclusion of `torch.func` support within `torch.compile` is currently under
+active development. At present, our support extends
+to the compilation of `grad` and `vmap`. However, it is important to note that
+there are certain [limitations](https://pytorch.org/docs/main/torch.compiler_faq.html#limitations)
+that restrict the range of cases we can compile.
+
+Looking ahead, our roadmap aims to extend the support for all transforms with
+minimal limitations, providing a more comprehensive compilation support for `torch.func`
+transforms
+
+## Closing Remarks
+This collective effort has been made achievable through the contributions of the
+PyTorch team at Quansight, as well as from Richard Zou and Horace He from Meta.
```

Comment:
Should we be more specific about what Quansight's involvement was here? It's not really clear reading this how much was done by us vs Meta people.

Reply:
Would rephrasing it as
> This collective effort has been made achievable through the contributions of the
PyTorch team at Quansight with help and guidance from Richard Zou and Horace He from Meta.

But it is worth mentioning that Richard and Horace added a lot of batch rules themselves and IIRC, Richard added the `operator` test for Composite Compliance.

Reply:
What about
> This project was yet another instance of the tight collaboration between Quansight and Meta within PyTorch. In particular, we would like to thank Richard Zou and Horace He, the functorch creators, for all the design discussions and guidance throughout these years.

Reply:
Will use this, thanks!